File: src/protocol/commands/image.ts
/**
 * Image upload payload builder.
 *
 * PNG data is split into 4096-byte chunks. Each chunk gets a 16-byte header
 * (same format as GIF upload):
 *   [0-1]:   u16 LE = header_len + chunk_data_len
 *   [2]:     0x01 (static)
 *   [3]:     0x00 (static)
 *   [4]:     flag (0 for first chunk, 2 for subsequent)
 *   [5-8]:   u32 LE = total PNG data length
 *   [9-12]:  u32 LE = CRC32 of entire PNG data
 *   [13-15]: 0x05, 0x00, 0x0D (static footer)
 *
 * Each chunk is sent separately with BLE write-with-response.
 */

import { IMAGE_CHUNK_SIZE } from "../constants.ts";
import { crc32 } from "../../utils/crc32.ts";

const HEADER_SIZE = 16;

export function setImageMode(mode: number): Uint8Array {
	return new Uint8Array([5, 0, 4, 1, mode & 0xff]);
}

export function createImagePayloads(pngData: Uint8Array): Uint8Array[] {
	const crc = crc32(pngData);
	const chunks = splitIntoChunks(pngData, IMAGE_CHUNK_SIZE);
	const payloads: Uint8Array[] = [];

	for (let i = 0; i < chunks.length; i++) {
		const chunk = chunks[i]!;
		const header = new Uint8Array(HEADER_SIZE);
		const view = new DataView(header.buffer);

		const chunkLen = HEADER_SIZE + chunk.length;
		view.setUint16(0, chunkLen, true);
		header[2] = 0x01;
		header[3] = 0x00;
		header[4] = i > 0 ? 0x02 : 0x00;
		view.setUint32(5, pngData.length, true);
		view.setUint32(9, crc, true);
		header[13] = 0x05;
		header[14] = 0x00;
		header[15] = 0x0d;

		const payload = new Uint8Array(chunkLen);
		payload.set(header, 0);
		payload.set(chunk, HEADER_SIZE);
		payloads.push(payload);
	}

	return payloads;
}

/** @deprecated Use createImagePayloads instead */
export function createImagePayload(pngData: Uint8Array): Uint8Array {
	const payloads = createImagePayloads(pngData);
	const totalLen = payloads.reduce((sum, p) => sum + p.length, 0);
	const result = new Uint8Array(totalLen);
	let offset = 0;
	for (const p of payloads) {
		result.set(p, offset);
		offset += p.length;
	}
	return result;
}

function splitIntoChunks(data: Uint8Array, chunkSize: number): Uint8Array[] {
	const chunks: Uint8Array[] = [];
	for (let i = 0; i < data.length; i += chunkSize) {
		chunks.push(data.slice(i, i + chunkSize));
	}
	return chunks;
}

File: rust-cli/src/protocol/commands/image.rs
#[allow(dead_code)]
use crate::protocol::constants::IMAGE_CHUNK_SIZE;

#[allow(dead_code)]
const HEADER_SIZE: usize = 16;

#[allow(dead_code)]
pub fn set_image_mode(mode: u8) -> Vec<u8> {
    vec![5, 0, 4, 1, mode]
}

#[allow(dead_code)]
pub fn create_image_payloads(png_data: &[u8]) -> Vec<Vec<u8>> {
    let crc = crc32fast::hash(png_data);
    let chunks: Vec<&[u8]> = png_data.chunks(IMAGE_CHUNK_SIZE).collect();
    let mut payloads = Vec::new();

    for (i, chunk) in chunks.iter().enumerate() {
        let chunk_len = HEADER_SIZE + chunk.len();
        let mut payload = vec![0u8; chunk_len];

        // Header
        payload[0..2].copy_from_slice(&(chunk_len as u16).to_le_bytes());
        payload[2] = 0x01;
        payload[3] = 0x00;
        payload[4] = if i > 0 { 0x02 } else { 0x00 };
        payload[5..9].copy_from_slice(&(png_data.len() as u32).to_le_bytes());
        payload[9..13].copy_from_slice(&crc.to_le_bytes());
        payload[13] = 0x05;
        payload[14] = 0x00;
        payload[15] = 0x0D;

        // Data
        payload[HEADER_SIZE..].copy_from_slice(chunk);
        payloads.push(payload);
    }

    payloads
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_set_image_mode() {
        assert_eq!(set_image_mode(1), vec![5, 0, 4, 1, 1]);
    }

    #[test]
    fn test_create_image_payloads_single_chunk() {
        let data = vec![0xAA; 100];
        let payloads = create_image_payloads(&data);
        assert_eq!(payloads.len(), 1);

        let p = &payloads[0];
        assert_eq!(p.len(), 116); // 16 + 100

        // Check header
        let len = u16::from_le_bytes([p[0], p[1]]);
        assert_eq!(len, 116);
        assert_eq!(p[2], 0x01);
        assert_eq!(p[3], 0x00);
        assert_eq!(p[4], 0x00); // first chunk
        let total_len = u32::from_le_bytes([p[5], p[6], p[7], p[8]]);
        assert_eq!(total_len, 100);
        assert_eq!(p[13], 0x05);
        assert_eq!(p[14], 0x00);
        assert_eq!(p[15], 0x0D);

        // Check data
        assert_eq!(&p[16..], &[0xAA; 100]);
    }

    #[test]
    fn test_create_image_payloads_multi_chunk() {
        let data = vec![0xBB; 5000]; // > 4096
        let payloads = create_image_payloads(&data);
        assert_eq!(payloads.len(), 2);

        // First chunk: flag=0
        assert_eq!(payloads[0][4], 0x00);
        assert_eq!(payloads[0].len(), 16 + 4096);

        // Second chunk: flag=2
        assert_eq!(payloads[1][4], 0x02);
        assert_eq!(payloads[1].len(), 16 + 904);
    }
}
